### ISSUE TITLE
Metadata v14 error during typegen

### ISSUE THREAD

**Issue by savudani8** [polkadot-js/api#4213]
I'm using `@polkadot/api` v6.8.1 and a chain with metadata v14 (`polkadot-v0.9.11`) whose metadata I retrieved in `parachain.json`. Not sure how to tackle this error on `polkadot-types-from-chain`
```
$ ts-node --skip-project node_modules/.bin/polkadot-types-from-chain --package @interlay/interbtc/interfaces --endpoint ./src/json/parachain.json --output ./src/interfaces
Generating from metadata, 145,369 bytes

interbtc-api/node_modules/@polkadot/types/codec/utils/decodeU8a.cjs:41
      throw new Error(`decodeU8a: failed at ${(0, _util.u8aToHex)(u8a.subarray(offset).slice(0, 8))}… on ${keys[i] ? `${keys[i]}` : ''}${rawType ? `: ${rawType}` : ''}:: ${error.message}`);
            ^
Error: decodeU8a: failed at 0x0e6104000c1c7370… on metadata:: decodeU8a: failed at 0x741853797374656d… on pallets: Vec<PalletMetadataV14>:: decodeU8a: failed at 0x1853797374656d01… on : {"name":"Text","storage":"Option<PalletStorageMetadataV14>","calls":"Option<PalletCallMetadataV14>","events":"Option<PalletEventMetadataV14>","constants":"Vec<PalletConstantMetadataV14>","errors":"Option<PalletErrorMetadataV14>","index":"u8"}:: decodeU8a: failed at 0x011853797374656d… on storage: Option<PalletStorageMetadataV14>:: decodeU8a: failed at 0x401c4163636f756e… on items: Vec<StorageEntryMetadataV14>:: decodeU8a: failed at 0x04e8205468652066… on : {"name":"Text","modifier":"StorageEntryModifierV14","type":"StorageEntryTypeV14","fallback":"Bytes","docs":"Vec<Text>"}:: decodeU8a: failed at 0x68652066756c6c20… on modifier: {"_enum":{"Optional":0,"Default":1,"Required":2}}:: Unable to create Enum via index 104, in Optional, Default, Required
    at decodeU8a (interbtc-api/node_modules/@polkadot/types/codec/utils/decodeU8a.cjs:41:13)
    at decodeStruct (interbtc-api/node_modules/@polkadot/types/codec/Struct.cjs:101:41)
    at new Struct (interbtc-api/node_modules/@polkadot/types/codec/Struct.cjs:135:26)
    at new MetadataVersioned (interbtc-api/node_modules/@polkadot/types/metadata/MetadataVersioned.cjs:51:5)
    at decodeMetadata (interbtc-api/node_modules/@polkadot/types/metadata/Metadata.cjs:30:14)
    at new Metadata (interbtc-api/node_modules/@polkadot/types/metadata/Metadata.cjs:55:21)
    at initMeta (interbtc-api/node_modules/@polkadot/typegen/util/initMeta.cjs:21:20)
    at generateDefaultConsts (interbtc-api/node_modules/@polkadot/typegen/generate/consts.cjs:94:27)
    at generate (interbtc-api/node_modules/@polkadot/typegen/fromChain.cjs:30:36)
    at main (interbtc-api/node_modules/@polkadot/typegen/fromChain.cjs:95:5)
```

**Comment by savudani8** [comment]
This is probably a bug introduced by `v6.8.1`, as typegen work with `v6.7.2`. However, it looks like in `6.7.2` there are some problems with determining the type names. For instance, some generated type are `InterbtcPrimitivesCurrencyId`, `InterbtcPrimitivesRedeemRedeemRequestStatus`, `InterbtcPrimitivesVaultId`, which should have probably been just `CurrencyId`, `RequestStatus`, `VaultId`. The metadata `path` parsing is probably at fault:
```
{
            "id": 36,
            "type": {
              "path": [
                "interbtc_primitives",
                "CurrencyId"
              ],
              "def": {
                "variant": {
                  "variants": [
                    {
                      "name": "DOT",
                      "index": 0
                    },
                    {
                      "name": "INTERBTC",
                      "index": 1
                    },
                    {
                      "name": "INTR",
                      "index": 2
                    },
                    {
                      "name": "KSM",
                      "index": 3
                    },
                    {
                      "name": "KBTC",
                      "index": 4
                    },
                    {
                      "name": "KINT",
                      "index": 5
                    }
                  ]
                }
              }
            }
          },
          ```

**Comment by jacogr** [comment]
The naming is correct. The names takes the full paths as specified in the metadata. (It does not map to the v13 names at all)

**Comment by savudani8** [comment]
I looked in a bit more detail at the generated files. Custom type imports are missing in the generated api files but they can all be found in `augmented-types.ts`. The problem that remains is that `augmented-types.ts` uses the v13 names (not the Java-like ones), so the same name parsing approach in `polkadot-types-from-chain` should be used in `polkadot-types-from-defs`

**Comment by savudani8** [comment]
After decoding the metadata, I think it should be possible to use `metadata.asLatest()` to get the `userDefs` used in `fromDefs` https://github.com/polkadot-js/api/blob/master/packages/typegen/src/fromDefs.ts#L29

Can you please provide me with some guidance to achieve that? I'll work on a PR where `polkadot-types-from-defs` would take an `--endpoint` input similar to `polkadot-types-from-chain` so it can use the metadata to generate a types file. Then it would be possible to import these types in the api interfaces from e.g. `'@interlay/interbtc/interfaces/default'`

**Comment by jacogr** [comment]
What seems to be missing is the lookup generation - https://github.com/polkadot-js/api/blob/master/packages/typegen/src/interfacesTs.ts#L7 (The API uses this script to build it's interfaces)

There is now basically 2 TS places -

- the old defs -> types, these are from user-generated types
- the lookup types, these are for the metadata types

So what seems to be the gap here is that none of the lookups are generated and outputted into a place form the scripts to pickup. So for the user-facing scripts, the lookup generation needs to be included - https://github.com/polkadot-js/api/blob/master/packages/typegen/src/generate/lookup.ts (it does both defs and then types from the metadata)

Since we can specify the output location as well as the source, it should be transparent? https://github.com/polkadot-js/api/blob/master/packages/typegen/src/generate/lookup.ts#L260

Now... the only there may be remaining is that when importing from lookup, it also takes the local paths into account. (Not 100% sure about this, underlying it uses all the same logic when doing txs, events, consts, so it should only be minor tweaks... hopefully?)

**Comment by jacogr** [comment]
Closed in #4224
